mmaction/apis/train.py

```diff
@@ -154,3 +161,65 @@ def train_model(model,
     if cfg.omnisource:
         runner_kwargs = dict(train_ratio=train_ratio)
     runner.run(data_loaders, cfg.workflow, cfg.total_epochs, **runner_kwargs)
+
+    if test['test_last'] or test['test_best']:
+        best_json = None
+        if test['test_best']:
+            best_json = osp.join(cfg.work_dir, 'best.json')
+            if not osp.exists(best_json):
+                test['test_best'] = False
+                runner.logger.info('Warning: test_best set as True, but is '
+                                   'not applicable')
+
+        if not (test['test_last'] or test['test_best']):
+            return
+
+        test_dataset = build_dataset(cfg.data.test, dict(test_mode=True))
+        gpu_collect = cfg.get('evaluation', {}).get('gpu_collect', False)
+        tmpdir = cfg.get('evaluation', {}).get('tmpdir',
+                                               osp.join(cfg.work_dir, 'tmp'))
+        dataloader_setting = dict(
+            videos_per_gpu=cfg.data.get('videos_per_gpu', 1),
+            workers_per_gpu=cfg.data.get('workers_per_gpu', 1),
+            num_gpus=len(cfg.gpu_ids),
+            dist=distributed,
+            shuffle=False)
+        dataloader_setting = dict(dataloader_setting,
+                                  **cfg.data.get('test_dataloader', {}))
+
+        test_dataloader = build_dataloader(test_dataset, **dataloader_setting)
+
+        names, ckpts = []
+
+        if test['test_last']:
+            names.append('last')
+            ckpts.append(None)
+        if test['test_best']:
+            assert best_json
+            best = mmcv.load(best_json)
+            best_ckpt = best['best_ckpt']
+            names.append(best)
+            ckpts.append(best_ckpt)
+
+        for name, ckpt in zip(names, ckpts):
+            if ckpt is not None:
+                runner.load_checkpoint(ckpt)
+
+            outputs = multi_gpu_test(runner.model, test_dataloader, tmpdir,
+                                     gpu_collect)
+            rank, _ = get_dist_info()
+            if rank == 0:
+                out = osp.join(cfg.work_dir, f'{name}_pred.pkl')
+                test_dataset.dump_results(outputs, out)
+
+                eval_cfg = cfg.get('evaluation', {})
+                for key in [
+                        'interval', 'tmpdir', 'start', 'gpu_collect',
+                        'save_best', 'rule', 'by_epoch', 'broadcast_bn_buffers'
+                ]:
+                    eval_cfg.pop(key, None)
+
+                eval_res = test_dataset.evaluate(outputs, **eval_cfg)
+                runner.logger.info('Testing results of the last checkpoint')
```

Comment:
May need to log best and last respectively

Reply:
Yes, I will modify here.